src/array_api_extra/_apply.py

```diff
@@ -0,0 +1,351 @@
+    dtypes : Sequence[DType], optional
+        Sequence of output dtypes, one for each output of `func`.
+        If `func` returns a single (non-sequence) output, this must be a sequence
+        with a single element.
+        Default: infer the result type(s) from the input arrays.
+    xp : array_namespace, optional
+        The standard-compatible namespace for `args`. Default: infer.
+    input_indices : Sequence[Sequence[Hashable]], optional
+        Dask specific.
+        Axes labels for each input array, e.g. if there are two args with respectively
+        ndim=3 and 1, `input_indices` could be ``['ijk', 'j']`` or ``[(0, 1, 2),
+        (1,)]``.
+        Default: disallow Dask.
+    core_indices : Sequence[Hashable], optional
+        **Dask specific.**
+        Axes of the input arrays that cannot be broken into chunks.
+        Default: disallow Dask.
+    output_indices : Sequence[Sequence[Hashable]], optional
+        **Dask specific.**
+        Axes labels for each output array. If `func` returns a single (non-sequence)
+        output, this must be a sequence containing a single sequence of labels, e.g.
+        ``['ijk']``.
+        Default: disallow Dask.
+    adjust_chunks : Sequence[Mapping[Hashable, Callable[[int], int]]], optional
+        **Dask specific.**
+        Sequence of dicts, one per output, mapping index to function to be applied to
+        each chunk to determine the output size. The total must add up to the output
+        shape.
+        Default: on Dask, the size along each index cannot change.
+    new_axes : Sequence[Mapping[Hashable, int]], optional
+        **Dask specific.**
+        New indexes and their dimension lengths, one per output.
+        Default: on Dask, there can't be `output_indices` that don't appear in
+        `input_indices`.
+    **kwargs : Any, optional
+        Additional keyword arguments to pass verbatim to `func`.
+        Any array objects in them won't be converted to NumPy.
+
+    Returns
+    -------
+    tuple[Array, ...]
+        The result(s) of `func` applied to the input arrays.
+        This is always a tuple, even if `func` returns a single output.
+
+    Notes
+    -----
+    JAX
+        This allows applying eager functions to jitted JAX arrays, which are lazy.
+        The function won't be applied until the JAX array is materialized.
+
+        The `JAX transfer guard
+        <https://jax.readthedocs.io/en/latest/transfer_guard.html>`_
+        may prevent arrays on a GPU device from being transferred back to CPU.
+        This is treated as an implicit transfer.
+
+    PyTorch, CuPy
+        These backends raise by default if you attempt to convert arrays on a GPU device
+        to NumPy.
+
+    Sparse
+        By default, sparse prevents implicit densification through ``np.asarray`.
+        `This safety mechanism can be disabled
+        <https://sparse.pydata.org/en/stable/operations.html#package-configuration>`_.
```

Comment:
Fine to leave as is for now I'd say. Once `sparse` adds a better API for this (an env var doesn't work), it seems reasonable to add a `force=False` option to this function. There are various reasons why one may want to force an expensive conversion; that kind of thing should always be opt-in on a case-by-case basis.

Reply:
Shouldn't this be tackled by a more general design pattern?

```python
with disable_sensible_guards():
    y = scipy.somefunc(x)
```
applies to torch/cupy/jax arrays on GPU, as well as sparse arrays.